designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala

```diff
@@ -66,83 +66,164 @@ class DeploymentServiceImpl(
-  // We split deploy process that way because we want to be able to split FE logic into two phases:
-  // - validations - it is quick part, the result will be displayed on deploy modal
-  // - deployment on engine side - it is longer part, the result will be shown as a notification
-  // TODO: refactor this into smaller steps, currently "validators" in this method have a lot of side effects and the whole thing is pretty confusing
+  /**
+    * Deploy is executed on the latest (current) scenario version.
+    * Inner Future in result allows to wait for deployment finish, while outer handles validation
+    * We split deploy process that way because we want to be able to split FE logic into two phases:
+    * - validations - it is quick part, the result will be displayed on deploy modal
+    * - deployment on engine side - it is longer part, the result will be shown as a notification
+    */
   override def deployProcessAsync(
-      processIdWithName: ProcessIdWithName,
+      processId: ProcessIdWithName,
       savepointPath: Option[String],
       comment: Option[String]
   )(implicit user: LoggedUser, ec: ExecutionContext): Future[Future[Option[ExternalDeploymentId]]] = {
     val actionType = ProcessActionType.Deploy
-    validateDeploymentComment(comment).flatMap { validatedComment =>
-      checkCanPerformActionAndAddInProgressAction[CanonicalProcess](
-        processIdWithName,
+
+    for {
+      // 1. common for all commands/actions
+      validatedComment <- validateDeploymentComment(comment)
+      ctx <- prepareCommandContextWithAction[CanonicalProcess](
+        processId,
         actionType,
-        d => Some(d.processVersionId),
-        d => Some(d.processingType)
-      ).flatMap { case (processDetails, actionId, versionOnWhichActionIsDone, buildInfoProcessIngType) =>
-        validateBeforeDeploy(processDetails, actionId).transformWith {
-          case Failure(ex) =>
-            dbioRunner.runInTransaction(actionRepository.removeAction(actionId)).transform(_ => Failure(ex))
-          case Success(validationResult) =>
-            // we notify of deployment finish/fail only if initial validation succeeded
-            val deploymentFuture = runDeploymentActionWithNotifications(
-              actionType,
-              actionId,
-              processIdWithName,
-              versionOnWhichActionIsDone,
-              validatedComment,
-              buildInfoProcessIngType
-            ) {
-              dispatcher
-                .deploymentManagerUnsafe(processDetails.processingType)
-                .processCommand(
-                  RunDeploymentCommand(
-                    validationResult.processVersion,
-                    validationResult.deploymentData,
-                    validationResult.resolvedScenario,
-                    savepointPath
-                  )
-                )
-            }
-            Future.successful(deploymentFuture)
-        }
+        getCurrentVersion,
```

Comment:
Please inline this function. It is not obvious which version is current as we use latest in most places in the code.